File: Cpp/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array.cpp
class Solution {
public:
    int searchFirstPosition(vector<int> &nums, int target){
        int lo = 0;
        int hi = nums.size();
        
        while(lo < hi){
            int mid = (lo + hi) / 2;
            if(nums[mid] < target) {
                lo = mid+1;
            }else {
                hi = mid;
            }
        }
        return lo;
    }
    
    int searchLastPosition(vector<int> &nums, int target){
        int lo = 0;
        int hi = nums.size();
        
        while(lo < hi){
            int mid = (lo + hi) / 2;
            if(nums[mid] <= target) {
                lo = mid+1;
            }else {
                hi = mid;
            }
        }
        
        return lo - 1;
    }
    
    vector<int> searchRange(vector<int>& nums, int target) {
        int first = searchFirstPosition(nums, target);
        int last = searchLastPosition(nums, target);
        
        if(first == nums.size() || nums[first] != target) {
            return {-1, -1};
        }
        
        return {first, last};
    }
};

File: Kotlin/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array.kt
class Solution {
    fun searchRange(nums: IntArray, target: Int): IntArray {
        val first = searchFirstPosition(nums, target)
        val last = searchLastPosition(nums, target)
        
        if(first == nums.size || nums[first] != target) {
            return intArrayOf(-1, -1)
        }
        
        return intArrayOf(first, last)
    }

    fun searchFirstPosition(nums: IntArray, target: Int): Int {
        var lo = 0
        var hi = nums.size

        while(lo < hi) {
            val mid = (lo + hi) / 2
            if(nums[mid] < target) {
                lo = mid + 1
            } else {
                hi = mid
            }
        }
        
        return lo
    }
    
    fun searchLastPosition(nums: IntArray, target: Int): Int {
        var lo = 0
        var hi = nums.size

        while (lo < hi) {
            val mid = (lo + hi) / 2
            if(nums[mid] <= target) {
                lo = mid + 1
            } else {
                hi = mid
            }
        }
        
        return lo - 1
    }
}